zboxcore/sdk/downloadworker.go

```diff
@@ -46,338 +45,494 @@ type DownloadRequest struct {
-				if len(req.encryptedKey) > 0 {
-
-					// dirty, but can't see other way right now
-					if req.authTicket == nil {
-						headerBytes := result.BlockChunks[blockNum][:EncryptionHeaderSize]
-						headerBytes = bytes.Trim(headerBytes, "\x00")
-
-						if len(headerBytes) != EncryptionHeaderSize {
-							logger.Logger.Error("Block has invalid header", req.blobbers[result.idx].Baseurl)
-							continue
-						}
-
-						encMsg := &encryption.EncryptedMessage{}
-						encMsg.EncryptedData = result.BlockChunks[blockNum][EncryptionHeaderSize:]
-
-						if len(headerBytes) != EncryptionHeaderSize {
-							logger.Logger.Error("Block has invalid header", req.blobbers[result.idx].Baseurl)
-							continue
-						}
-						encMsg.MessageChecksum, encMsg.OverallChecksum = string(headerBytes[:128]), string(headerBytes[128:])
-						encMsg.EncryptedKey = encscheme.GetEncryptedKey()
-						decryptedBytes, err := encscheme.Decrypt(encMsg)
-						if err != nil {
-							logger.Logger.Error("Block decryption failed", req.blobbers[result.idx].Baseurl, err)
-							continue
-						}
-						shards[blockNum][result.idx] = decryptedBytes
-					} else {
-						suite := edwards25519.NewBlakeSHA256Ed25519()
-						reEncMessage := &encryption.ReEncryptedMessage{
-							D1: suite.Point(),
-							D4: suite.Point(),
-							D5: suite.Point(),
-						}
-						err := reEncMessage.Unmarshal(result.BlockChunks[blockNum])
-						if err != nil {
-							logger.Logger.Error("ReEncrypted Block unmarshall failed", req.blobbers[result.idx].Baseurl, err)
-							break
-						}
-						decrypted, err := encscheme.ReDecrypt(reEncMessage)
-						if err != nil {
-							logger.Logger.Error("Block redecryption failed", req.blobbers[result.idx].Baseurl, err)
-							break
-						}
-						shards[blockNum][result.idx] = decrypted
-					}
-				} else {
-					shards[blockNum][result.idx] = result.BlockChunks[blockNum]
-				}
-
-				// All share should have equal length
-				decodeLen[blockNum] = len(shards[blockNum][result.idx])
-				blockSuccess = true
-			}
+	}
+	return data, nil
+}
 
-			if !blockSuccess {
-				continue
-			}
+// downloadBlock This function will add download requests to the download channel which picks up
+// download requests and processes it.
+// This function will fill up `shards` in respective position and also return failed number of
+// blobbers along with remainingMask that are the blobbers that are not yet requested.
+func (req *DownloadRequest) downloadBlock(
+	startBlock, totalBlock int64,
+	mask zboxutil.Uint128, requiredDownloads int,
+	shards [][][]byte) (zboxutil.Uint128, int, error) {
+
+	var remainingMask zboxutil.Uint128
+	activeBlobbers := mask.CountOnes()
+	if activeBlobbers < requiredDownloads {
+		return zboxutil.NewUint128(0), 0, errors.New("insufficient_blobbers",
+			fmt.Sprintf("Required downloads %d, remaining active blobber %d",
+				req.consensusThresh, activeBlobbers))
+	}
+	rspCh := make(chan *downloadBlock, req.consensusThresh)
```

Comment:
Actually, we don't request more than consensusThresh. It will suffice. 
`requiredDownload` is always lesser than consensusThresh. So its enough to use `requiredDownload` instead of consensusThresh